Python/Product/Analysis/LanguageServer/Messages.cs

```diff
@@ -0,0 +1,302 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+
+namespace Microsoft.PythonTools.Analysis.LanguageServer {
+    public struct InitializeParams {
+        public int? processId;
+        public string rootPath;
+        public Uri rootUri;
+        public PythonInitializationOptions initializationOptions;
+        public ClientCapabilities? capabilities;
+        public TraceLevel trace;
+    }
+
+    [Serializable]
+    public class LanguageServerException : Exception {
+        public const int UnknownDocument = 1;
+        public const int UnsupportedDocumentType = 2;
+        public const int MismatchedVersion = 3;
+
+        public int Code => (int)Data["Code"];
+
+        public sealed override System.Collections.IDictionary Data => base.Data;
+
+        public LanguageServerException(int code, string message) : base(message) {
+            Data["Code"] = code;
+        }
+
+        public LanguageServerException(int code, string message, Exception innerException) : base(message, innerException) {
+            Data["Code"] = code;
+        }
+    }
+
+    public struct InitializeResult {
+        public ServerCapabilities? capabilities;
```

Comment:
`struct` with mutable field. `capabilities` can be made `readonly` and initialized from constructor

Reply:
These are the public interface for the LSP implementation and I want them to be as close as possible to JSON objects for ease of mapping.

Reply:
Ok, got it.